File: tareas/2/24190108/Lista 17/Cadenas Buenas.py
def invertir_cadena(s):
    return s[::-1]

def convertir_a_buena(s):
    caracteres = []
    for char in s:
        if caracteres and caracteres[-1].islower() and caracteres[-1].upper() == char:
            caracteres.pop()
        else:
            caracteres.append(char)

    return invertir_cadena(''.join(caracteres))

if __name__ == "__main__":
    print(convertir_a_buena("UNMSM"))
    print(convertir_a_buena("sS"))
    print(convertir_a_buena("xaBBAyCcZ"))
    print(convertir_a_buena("xaaabbbBBBBAAAAd"))

File: ejercicios/19_20_21_22/cadenas_buenas/cadenas_buenas.cpp
#include <iostream>
#include <stack>
using namespace std;

string InvertirCadena(string s) {
	string r;
	for (int i = s.size() - 1; i >= 0; i--) {
		r += s[i];
	}
	return r;
}

string ConvertirABuena(string s) {
	stack<char> caracteres;
	for (int i = 0; i < s.size(); i++) {
		if (
			!caracteres.empty() &&
			islower(caracteres.top()) &&
			toupper(caracteres.top()) == s[i]
		) {
			caracteres.pop();
		} else {
			caracteres.push(s[i]);
		}
	}
	
	string r;
	while (!caracteres.empty()) {
		r += caracteres.top();
		caracteres.pop();
	}
	
	return InvertirCadena(r);
}

int main(void) {
	cout << ConvertirABuena("UNMSM") << endl;
	cout << ConvertirABuena("s") << endl;
	cout << ConvertirABuena("xabBAycCz") << endl;
	cout << ConvertirABuena("xaaabbbBBBAAAd") << endl;
	return 0;
}